File: web/src/features/selection/batchEdit.ts
import { TIME_SIGNATURES, type Instrument } from '../../api/vocabulary'
import type { BulkPatch } from '../../commands/bulk'
import { STATUS_LABELS } from '../../constants'
import type { CatalogEntry } from '../catalog/filters'
import { isTuneStatus } from '../catalog/status'
import {
  byTuningKey,
  isTuningKey,
  TUNING_KEYS,
  tuningEntry,
  tuningKeyInstrument,
  tuningLabel,
} from '../settings/instruments'
import { DETAIL_LABELS } from '../tune/detailFields'
import { isMode } from '../tune/keyMode'

export const EDIT_FIELDS = [
  'status',
  'key',
  'mode',
  ...TUNING_KEYS,
  'genre',
  'tune_type',
  'time_signature',
  'part_structure',
  'is_crooked',
  'learned_from',
  'learned_on',
] as const

export type EditField = (typeof EDIT_FIELDS)[number]

/** The instrument whose tuning a field edits, or undefined for a column. */
export function tuningInstrument(field: EditField): Instrument | undefined {
  return tuningKeyInstrument(field)
}

export const EDIT_FIELD_LABELS: Record<EditField, string> = {
  status: 'Status',
  key: 'Key',
  mode: DETAIL_LABELS.mode,
  ...byTuningKey(tuningLabel),
  genre: DETAIL_LABELS.genre,
  tune_type: DETAIL_LABELS.tune_type,
  time_signature: DETAIL_LABELS.time_signature,
  part_structure: DETAIL_LABELS.part_structure,
  is_crooked: DETAIL_LABELS.is_crooked,
  learned_from: DETAIL_LABELS.learned_from,
  learned_on: DETAIL_LABELS.learned_on,
}

/** A choice field has a vocabulary and is picked from a list; text and date fields are typed. */
export const FIELD_KINDS: Record<EditField, 'choice' | 'text' | 'date' | 'boolean'> = {
  status: 'choice',
  key: 'choice',
  mode: 'choice',
  ...byTuningKey(() => 'choice' as const),
  genre: 'choice',
  tune_type: 'choice',
  time_signature: 'choice',
  part_structure: 'choice',
  is_crooked: 'boolean',
  learned_from: 'text',
  learned_on: 'date',
}

const USER_TUNE_FIELDS = ['status', 'learned_from', 'learned_on'] as const satisfies EditField[]

type UserTuneField = (typeof USER_TUNE_FIELDS)[number]

function isUserTuneField(field: EditField): field is UserTuneField {
  return USER_TUNE_FIELDS.some((userField) => userField === field)
}

export type Summary =
  { kind: 'shared'; value: string | boolean } | { kind: 'mixed' } | { kind: 'empty' }

/** A string for text, select, and date fields, a boolean for yes or no, and null to clear. */
export type TouchedValue = string | boolean | null
export type Touched = Partial<Record<EditField, TouchedValue>>

// A server row can carry a value from a schema version this client predates;
// fall back to empty rather than trust it as one of this client's known options.
function fieldValue(entry: CatalogEntry, field: EditField): string | boolean | null {
  if (isTuningKey(field)) {
    const instrument = tuningKeyInstrument(field)
    return instrument ? tuningEntry(entry.tune.tunings, instrument).tuning : null
  }
  if (field === 'mode') {
    const modes = entry.tune.modes.filter(isMode)
    return modes.length > 0 ? modes.join(', ') : null
  }
  const value: unknown = isUserTuneField(field) ? entry.userTune[field] : entry.tune[field]
  if (typeof value !== 'string' && typeof value !== 'boolean') return null
  if (field === 'time_signature')
    return TIME_SIGNATURES.some((signature) => signature === value) ? value : null
  if (field === 'status') return typeof value === 'string' && isTuneStatus(value) ? value : null
  return value
}

export function summarize(entries: readonly CatalogEntry[]): Record<EditField, Summary> {
  const summaries = {} as Record<EditField, Summary>
  for (const field of EDIT_FIELDS) {
    const values = entries.map((entry) => fieldValue(entry, field))
    const first = values[0] ?? null
    if (!values.every((value) => value === first)) summaries[field] = { kind: 'mixed' }
    else summaries[field] = first === null ? { kind: 'empty' } : { kind: 'shared', value: first }
  }
  return summaries
}

/** Every field, except a tuning for an instrument the user does not play that no selected tune fills. */
export function visibleEditFields(
  entries: readonly CatalogEntry[],
  instruments: ReadonlySet<Instrument>,
): EditField[] {
  return EDIT_FIELDS.filter((field) => {
    const instrument = tuningInstrument(field)
    if (!instrument) return true
    return (
      instruments.has(instrument) ||
      entries.some((entry) => tuningEntry(entry.tune.tunings, instrument).tuning !== null)
    )
  })
}

function normalize(value: TouchedValue): string | boolean | null {
  return typeof value === 'string' ? value.trim() || null : value
}

/**
 * True when saving the value would leave every selected tune as it is. Compares the
 * raw value, not the trimmed one that is saved, so a trailing space typed mid-edit does
 * not snap a text field back to untouched.
 */
export function isUnchanged(summary: Summary, value: TouchedValue): boolean {
  if (value === null || value === '') return summary.kind === 'empty'
  return summary.kind === 'shared' && summary.value === value
}

export function toPatch(touched: Touched): BulkPatch {
  const tune: Record<string, unknown> = {}
  const userTune: Record<string, unknown> = {}
  const tunings: Partial<Record<Instrument, string | null>> = {}
  for (const field of EDIT_FIELDS) {
    const raw = touched[field]
    if (raw === undefined) continue
    const value = normalize(raw)
    if (field === 'status' && value === null) continue
    if (field === 'mode') {
      tune.modes = typeof value === 'string' ? [value] : []
      continue
    }
    const instrument = tuningInstrument(field)
    if (instrument) tunings[instrument] = typeof value === 'string' ? value : null
    else (isUserTuneField(field) ? userTune : tune)[field] = value
  }
  const patch: BulkPatch = {
    tune: tune as BulkPatch['tune'],
    userTune: userTune as BulkPatch['userTune'],
  }
  if (Object.keys(tunings).length > 0) patch.tunings = tunings
  return patch
}

export function displayValue(field: EditField, value: string | boolean): string {
  if (typeof value === 'boolean') return value ? 'yes' : 'no'
  if (field === 'status' && isTuneStatus(value)) return STATUS_LABELS[value]
  return value
}

File: apple/CrosstuneKit/Sources/CrosstuneUI/Selection/BatchEdit.swift
import CrosstuneCommands
import CrosstuneStore
import CrosstuneVocabulary
import Foundation

/// A field the bulk edit sheet sets across many tunes. A field whose value belongs to one tune
/// alone, such as a title, a composer, or notes, is never here.
public enum EditField: Hashable, Sendable {
    case status
    case key
    case mode
    case tuning(String)
    case genre
    case tuneType
    case timeSignature
    case partStructure
    case isCrooked
    case learnedFrom
    case learnedOn

    /// How a field is set: picked from a list, typed, picked as a day, or yes or no.
    public enum Kind: Sendable {
        case choice
        case text
        case date
        case yesNo
    }

    /// Every field, in the order the sheet reads.
    public static let all: [EditField] =
        [.status, .key, .mode] + Vocabulary.instruments.map(EditField.tuning)
        + [.genre, .tuneType, .timeSignature, .partStructure, .isCrooked, .learnedFrom, .learnedOn]

    /// The field's name, as the tune form calls it.
    public var label: String {
        switch self {
        case .status: TuneFieldLabels.status
        case .key: TuneFieldLabels.key
        case .mode: TuneFieldLabels.partMode(0)
        case .tuning(let instrument): TuneFieldLabels.tuning(instrument)
        case .genre: TuneFieldLabels.genre
        case .tuneType: TuneFieldLabels.tuneType
        case .timeSignature: TuneFieldLabels.timeSignature
        case .partStructure: TuneFieldLabels.partStructure
        case .isCrooked: TuneFieldLabels.isCrooked
        case .learnedFrom: TuneFieldLabels.learnedFrom
        case .learnedOn: TuneFieldLabels.learnedOn
        }
    }

    public var kind: Kind {
        switch self {
        case .isCrooked: .yesNo
        case .learnedFrom: .text
        case .learnedOn: .date
        default: .choice
        }
    }

    /// The instrument whose tuning this field sets, or nil for a column.
    public var instrument: String? {
        if case .tuning(let instrument) = self { return instrument }
        return nil
    }
}

/// One field's value on one tune: text for a text, choice, or date field, a flag for yes or no.
public enum EditValue: Hashable, Sendable {
    case text(String)
    case flag(Bool)
}

/// What the selected tunes hold in one field.
public enum EditSummary: Hashable, Sendable {
    /// Every tune holds this value.
    case shared(EditValue)
    /// The tunes disagree.
    case mixed
    /// No tune holds a value.
    case empty
}

/// A field the musician has touched: a value to write, or clear to empty it on every tune.
public enum TouchedValue: Hashable, Sendable {
    case text(String)
    case flag(Bool)
    case clear
}

/// The bulk edit sheet's rules: what the selected tunes share, which fields it offers, when a
/// touch changes nothing, and the one write a save makes.
public enum BatchEdit {
    /// The value `entry` holds in `field`, or nil. A value from a newer server that this build
    /// does not know reads as none, rather than being trusted as one of the known choices.
    static func value(_ entry: CatalogEntry, _ field: EditField) -> EditValue? {
        let tune = entry.tune
        let userTune = entry.userTune
        switch field {
        case .status:
            return Vocabulary.statusLabels[userTune.status] == nil ? nil : .text(userTune.status)
        case .key: return tune.key.map(EditValue.text)
        case .mode:
            let modes = tune.modes.filter(Vocabulary.modes.contains)
            return modes.isEmpty ? nil : .text(modes.joined(separator: ", "))
        case .tuning(let instrument):
            return tuningEntry(tune.tunings, instrument: instrument).tuning.map(EditValue.text)
        case .genre: return tune.genre.map(EditValue.text)
        case .tuneType: return tune.tuneType.map(EditValue.text)
        case .timeSignature:
            guard let signature = tune.timeSignature, Vocabulary.timeSignatures.contains(signature) else { return nil }
            return .text(signature)
        case .partStructure: return tune.partStructure.map(EditValue.text)
        case .isCrooked: return .flag(tune.isCrooked)
        case .learnedFrom: return userTune.learnedFrom.map(EditValue.text)
        case .learnedOn: return userTune.learnedOn.map(EditValue.text)
        }
    }

    /// What the tunes share in every field.
    public static func summarize(_ entries: [CatalogEntry]) -> [EditField: EditSummary] {
        var summaries: [EditField: EditSummary] = [:]
        for field in EditField.all {
            let values = entries.map { value($0, field) }
            let first = values.first ?? nil
            if !values.allSatisfy({ $0 == first }) {
                summaries[field] = .mixed
            } else {
                summaries[field] = first.map(EditSummary.shared) ?? .empty
            }
        }
        return summaries
    }

    /// Every field, less a tuning for an instrument the musician does not play and no selected
    /// tune has a tuning for. A capo alone does not count.
    public static func visibleFields(_ entries: [CatalogEntry], instruments: Set<String>) -> [EditField] {
        EditField.all.filter { field in
            guard let instrument = field.instrument else { return true }
            return instruments.contains(instrument)
                || entries.contains { tuningEntry($0.tune.tunings, instrument: instrument).tuning != nil }
        }
    }

    /// True when saving `value` would leave every tune as it is. Compares the raw text, not the
    /// trimmed text a save writes, so a space typed mid-word does not snap the row back to
    /// untouched.
    public static func isUnchanged(_ summary: EditSummary, _ value: TouchedValue) -> Bool {
        switch value {
        case .clear, .text(""): summary == .empty
        case .text(let text): summary == .shared(.text(text))
        case .flag(let flag): summary == .shared(.flag(flag))
        }
    }

    /// The write for the touched fields. A value outside its field's vocabulary is skipped, so
    /// it never reaches the outbox, and status is never cleared.
    public static func patch(_ touched: [EditField: TouchedValue]) -> BulkPatch {
        var patch = BulkPatch()
        for field in EditField.all {
            guard let raw = touched[field] else { continue }
            let text: String?
            let flag: Bool?
            switch raw {
            case .text(let value):
                let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
                text = trimmed.isEmpty ? nil : trimmed
                flag = nil
            case .flag(let value):
                text = nil
                flag = value
            case .clear:
                text = nil
                flag = nil
            }
            switch field {
            case .status:
                if let text, Vocabulary.statuses.contains(text) { patch.userTune.status = .value(text) }
            case .learnedFrom: patch.userTune.learnedFrom = .value(text)
            case .learnedOn: patch.userTune.learnedOn = .value(text)
            case .mode:
                if let text {
                    if Vocabulary.modes.contains(text) { patch.tune.modes = .value([text]) }
                } else {
                    patch.tune.modes = .value([])
                }
            case .timeSignature:
                if text.map(Vocabulary.timeSignatures.contains) ?? true { patch.tune.timeSignature = .value(text) }
            case .isCrooked:
                if let flag { patch.tune.isCrooked = .value(flag) }
            case .key: patch.tune.key = .value(text)
            case .genre: patch.tune.genre = .value(text)
            case .tuneType: patch.tune.tuneType = .value(text)
            case .partStructure: patch.tune.partStructure = .value(text)
            case .tuning(let instrument): patch.tunings[instrument] = .value(text)
            }
        }
        return patch
    }
}

/// The bulk edit sheet's state: what the selected tunes share and what the musician has
/// touched. Only a touched field is written.
public struct BulkEditForm: Equatable, Sendable {
    public static let mixed = "Mixed"
    /// A choice row's empty choice, which empties the field on every tune.
    public static let clear = "Clear"
    /// A yes or no row's empty choice, which leaves every tune as it is.
    public static let keep = "Keep"
    public static let yes = "Yes"
    public static let no = "No"
    public static let footnote = "Only fields you change are saved."

    public let summaries: [EditField: EditSummary]
    /// The fields offered, in order.
    public let fields: [EditField]
    public private(set) var touched: [EditField: TouchedValue] = [:]

    public init(entries: [CatalogEntry], instruments: Set<String>) {
        summaries = BatchEdit.summarize(entries)
        fields = BatchEdit.visibleFields(entries, instruments: instruments)
    }

    public func summary(_ field: EditField) -> EditSummary {
        summaries[field] ?? .empty
    }

    /// Sets a field, or leaves it untouched when `value` is nil or would change no tune.
    public mutating func touch(_ field: EditField, _ value: TouchedValue?) {
        if let value, !BatchEdit.isUnchanged(summary(field), value) {
            touched[field] = value
        } else {
            touched[field] = nil
        }
    }

    /// Whether anything would be written, which is what makes Save live.
    public var isEdited: Bool { !touched.isEmpty }

    /// The save's write.
    public var patch: BulkPatch { BatchEdit.patch(touched) }

    /// The row's value: the touched one, else the one every tune shares, else nil.
    public func value(_ field: EditField) -> EditValue? {
        switch touched[field] {
        case .text(let text): return text.isEmpty ? nil : .text(text)
        case .flag(let flag): return .flag(flag)
        case .clear: return nil
        case nil:
            if case .shared(let value) = summary(field) { return value }
            return nil
        }
    }

    /// The row's value as text, empty when it has none.
    public func text(_ field: EditField) -> String {
        if case .text(let text) = value(field) { return text }
        return ""
    }

    /// Whether the row still stands for tunes that disagree.
    public func isMixed(_ field: EditField) -> Bool {
        touched[field] == nil && summary(field) == .mixed
    }

    /// What the row says when it holds no value: Mixed for tunes that disagree, else Not set.
    public func placeholder(_ field: EditField) -> String {
        isMixed(field) ? Self.mixed : TuneFieldLabels.notSet
    }

    /// What the row shows: its value, Not set, or Mixed.
    public func shown(_ field: EditField) -> String {
        switch value(field) {
        case .text(let text): field == .status ? StatusStyle.label(text) : text
        case .flag(let flag): flag ? Self.yes : Self.no
        case nil: placeholder(field)
        }
    }
}
